lib/api/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+        FuncSig, GlobalDescriptor, MemoryDescriptor, TableDescriptor, Type, Value,
+    };
+}
+
+pub mod import {
+    //! Types and functions for Wasm imports.
+    pub use wasmer_runtime_core::module::{Import, ImportType};
+    pub use wasmer_runtime_core::{func, imports};
+}
+
+pub mod export {
+    //! Types and functions for Wasm exports.
+    pub use wasmer_runtime_core::module::{ExportDescriptor, ExportKind};
+}
+
+pub mod units {
+    //! Various unit types.
+    pub use wasmer_runtime_core::units::{Bytes, Pages};
+}
+
+pub mod types {
+    //! Types used in the Wasm runtime and conversion functions.
+    pub use wasmer_runtime_core::types::*;
+}
+
+pub mod error {
+    //! Various error types returned by Wasmer APIs.
+    pub use wasmer_runtime_core::error::{CompileError, CompileResult};
+}
+
+pub use prelude::*;
+
+/// Idea for generic trait; consider rename; it will need to be moved somewhere else
+pub trait CompiledModule {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary_unchecked(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_file(file: impl AsRef<std::path::Path>) -> error::CompileResult<Module>;
+
+    fn validate(bytes: impl AsRef<[u8]>) -> error::CompileResult<()>;
+}
+
+use wasmer_runtime_core::backend::Compiler;
+
+/// Copied from runtime core; TODO: figure out what we want to do here
+pub fn default_compiler() -> impl Compiler {
+    #[cfg(any(
+        all(
+            feature = "default-backend-llvm",
+            not(feature = "docs"),
+            any(
+                feature = "default-backend-cranelift",
+                feature = "default-backend-singlepass"
+            )
+        ),
+        all(
+            not(feature = "docs"),
+            feature = "default-backend-cranelift",
+            feature = "default-backend-singlepass"
+        )
+    ))]
+    compile_error!(
+        "The `default-backend-X` features are mutually exclusive.  Please choose just one"
+    );
+
+    #[cfg(all(feature = "default-backend-llvm", not(feature = "docs")))]
+    use wasmer_llvm_backend::LLVMCompiler as DefaultCompiler;
+
+    #[cfg(all(feature = "default-backend-singlepass", not(feature = "docs")))]
+    use wasmer_singlepass_backend::SinglePassCompiler as DefaultCompiler;
+
+    #[cfg(any(feature = "default-backend-cranelift", feature = "docs"))]
+    use wasmer_clif_backend::CraneliftCompiler as DefaultCompiler;
+
+    DefaultCompiler::new()
+}
+
+// this implementation should be moved
+impl CompiledModule for Module {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        let bytes = bytes.as_ref();
+        wasmer_runtime_core::compile_with(bytes, &default_compiler())
+    }
+
+    fn from_binary(_bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        todo!("from_binary: how is this different from `new`?")
```

Comment:
Implementation missing :)

Note: `new` calls from_binary. But perhaps we want to add automatic support for converting text files into wasm (in `new` not in `from_binary`).